components/zcash_note_encryption/src/batch.rs

```diff
@@ -55,17 +65,18 @@ where
     let keys = D::batch_kdf(items);
 
     // Finish the trial decryption!
-    ivks.iter()
-        .flat_map(|ivk| {
-            // Reconstruct the matrix of (ivk, output) combinations.
-            iter::repeat(ivk)
-                .zip(ephemeral_keys.iter())
-                .zip(outputs.iter())
-        })
-        .zip(keys)
-        .map(|(((ivk, (_, ephemeral_key)), (domain, output)), key)| {
-            // The `and_then` propagates any potential rejection from `D::epk`.
-            key.and_then(|key| decrypt_inner(domain, ivk, ephemeral_key, output, key))
+    keys.chunks(ivks.len())
+        .zip(ephemeral_keys.iter().zip(outputs.iter()))
+        .map(|(key_chunk, ((_, ephemeral_key), (domain, output)))| {
+            key_chunk
+                .iter()
+                .zip(ivks.iter().enumerate())
+                .filter_map(|(key, (i, ivk))| {
+                    key.as_ref()
+                        .and_then(|key| decrypt_inner(domain, ivk, ephemeral_key, output, key))
+                        .map(|out| (out, i))
+                })
+                .next()
```

Comment:
The commit message claims that `m * n` decryptions is only the worst case, and this change to return early results in a 50% performance improvement to multi-key scanning on average. I disagree on both counts. `m * n` decryptions is indeed the worst case, but it is also the average case: the chain will contain many more outputs that are _not_ for the wallet than it does that _are_ for the wallet. The 50% improvement only applies to the cost of decrypting notes that are for us; it is not a 50% improvement to the average cost of scanning. The commit text needs to be fixed to clarify this.

Separately, a downside to making this return early is that trial decryption cannot be constant time with respect to the IVKs (by definition). This should be noted in the documentation. I think it is currently not constant time with respect to _outputs_ because the post-decryption validation steps in `zcash_note_encryption` have early returns, but that's a separate issue. cc @defuse, who was the primary motivator for closer-to-constant-time detection when the early mobile SDK code was built, resulting in code like this:

https://github.com/zcash/librustzcash/blob/37fc28634e811fdf8db9274a0080cff17c6b6a09/zcash_client_backend/src/welding_rig.rs#L238-L265).